util/src/graph.cpp

```diff
@@ -68,81 +68,87 @@ int Graph::edgeCount() const
   return agnedges(_graphPimpl->_graph);
 }
 
-Graph::Node Graph::addNode(const Subgraph& subgraph_)
+Graph::Node Graph::createNode(const Subgraph& subgraph_)
 {
-  std::string id = generateId();
+  return getOrCreateNode(generateId(), subgraph_);
+}
+
+Graph::Node Graph::getOrCreateNode(
+  const Graph::Node& id_,
+  const Subgraph& subgraph_)
+{
+  std::string id = id_.empty() ? generateId() : id_;
+  _ids.insert(id);
 
   agnode(
-    subgraph_.id.empty()
+    subgraph_.empty()
       ? _graphPimpl->_graph
-      : _graphPimpl->_subgMap[subgraph_.id],
+      : _graphPimpl->_subgMap[subgraph_],
     const_cast<char*>(id.c_str()),
     1);
 
-  Node node;
-  node.id = id;
-  setAttribute(node, "fontsize", "11");
+  setNodeAttribute(id_, "fontsize", "11");
```

Comment:
This `getOrCreateNode()` function should also set the `id` attribute of the node.